idseq_dag/util/s3.py

```diff
@@ -82,6 +83,39 @@ def check_s3_presence(s3_path, allow_zero_byte_files=True):
         return _check_s3_presence(s3_path, allow_zero_byte_files)
 
 
+def _list_s3_keys(s3_path_prefix):
+    """Returns an iterator of s3 keys prefixed by s3_path_prefix."""
+    with log.log_context(context_name="s3.list_s3_objects", values={'s3_path_prefix': s3_path_prefix}, log_context_mode=log.LogContextMode.EXEC_LOG_EVENT):
+        parsed_url = urlparse(s3_path_prefix, allow_fragments=False)
+        bucket = parsed_url.netloc
+        prefix = parsed_url.path.lstrip('/')
+        client = boto3.client('s3')
+        paginator = client.get_paginator('list_objects')
+        for response in paginator.paginate(Bucket=bucket, Prefix=prefix):
+            for item in response['Contents']:
+                yield item['Key']
+
+
+def list_s3_keys(s3_path_prefix):
+    with botolock:
+        rate_limit_boto()
+        return _list_s3_keys(s3_path_prefix)
```

Comment:
Sorry I am late to the party here but I have a number of concerns with the use of boto3 regarding the fact that it tends to implicitly use a global session object that cannot be shared between threads.  

 Is there a way to rewrite this without using boto3?   If not, would you mind creating your own dedicated boto session for every client ant paginator creation?

You could look for instance here https://stackoverflow.com/questions/52820971/is-boto3-client-thread-safe but by no means is that the only place where its thread safety is questioned.  The concern that the implicit global boto session should not be used from multiple threads is documented in the official boto docs.   All code in s3.py is meant to be thread safe, and this example is even more important because the paginator is a long lived object (a generator).

If you must use boto3, it seems to me the botolock and the rate_limit_boto() functions are placed in the wrong place.  Those should protect lines 92 and 93 above, where the first boto3 operations occur, not this line.  It's also unclear to me if the paginator